lib/platform/linux-uart.c

```diff
@@ -0,0 +1,437 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+const char *dbginfo_off = "pscdbg 0 all\r";
+const char *echo_off = "echo 0\r";
+const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(50)
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	int idx;
+	int rty_cnt = 0;
+	char rtn[1024];
+	int cnt = 0;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = write(udev->fd, str, strlen(str));
+	if (ret != strlen(str))
+		return -1;
+
+cli_rtn_retry:
+	if (rty_cnt > RETRY_NUM)
+		return -1;
+
+	ret = read(udev->fd, rtn + cnt, sizeof(rtn));
+	if (ret < 0)
+		return ret;
+
+	cnt += ret;
+	rtn[cnt] = '\0';
+
+	if (sscanf(rtn, "%*[^:]:%x>", &idx) != 1){
+		usleep(5000);
```

Comment:
Actually, current usleep(5000) and RETRY_NUM (now 50) is a combination results after adjusting/tuning with real case.
For example for fw-toggle, which take very long time waiting for the response for the toggle MRPC command.
If remove the usleep, then the RETRY_NUM will increased proportionly, and more syscall read will invoked.
In real case, once the usleep is called, it means that the response is not fully collect from device, it is a punishment for this utility, to decrease its calling of syscall read.

Reply:
Then it would be better to remove the usleep and increase the timeout on the read command to match.

Reply:
I guess you would like FW to implement the GAS READ COMMD "gasrd -c -s <offset> [byte count]" with an parameter "-t timeout", right?
Current FW have no this timeout parameter for gasrd cmd.

Reply:
No, that doesn't seem necessary. Just make sure the timeout has a sensible value.

Also, the FW has nothing to do with the timeout.  The timeout is a feature of the OS (ie. how long read() will wait with no data before returning).

Reply:
Hi, Logan,

Actually, we have following set for the termios:
 
+	uart_attribs.c_cc[VMIN] = 0;  
+	uart_attribs.c_cc[VTIME] = 5; 

MIN == 0; TIME > 0: TIME specifies the limit for a timer in tenths of a second. The timer is started when read(2) is called. read(2) returns either when at least one byte of data is available, or when the timer expires. If the timer expires without any input becoming available, read(2) returns 0. 

But it does not help for our case:
because FW response to the gasrd/wr command in the form of strings, one by one.
from timeout aspect, it doesn't timeout, because FW keep sending strings.
but from the entire response from fw, maybe the response with long payload, need several read+usleep to collect all the response.

Reply:
Ok, that's what I thought. But then what you are saying about needing the usleep() still makes no sense. Waiting for 5ms + 0.5s vs only 0.5s shouldn't change anything significantly as you're affecting the delay between reads by 1%.

The things you talk about the firmware also makes no sense. You keep reading lines until one times out or you've read all expected data (you may also want a longer timeout as it should only timeout when something goes wrong). Sleeping between the reads doesn't help and you probably don't even need a retry counter.  If it times out waiting for data you expect, that should signal something bad happened and there's a problem, no need to do additional read() calls.

Reply:
Sleeping between the reads doesn't help and you probably don't even need a retry counter. 

sleeping between the reads is reduce the read syscall.
the point is, when fw response in flight, the read will not timeout, remove the usleep, result in more syscall read needed.

Reply:
Sleeping between the reads doesn't help and you probably don't even need a retry counter.
sleeping between the reads is reduce the read syscall.
the point is, when fw response in flight, the read will not timeout, remove the usleep, result in more syscall read needed.

Reply:
That makes absolutely no sense.

Your are either waiting wastefully in a sleep syscall or productively in a read syscall. The read call will block until there's something useful to return. So the extra sleep is unnecessary.